src/toil/mpcopy.py

```diff
@@ -0,0 +1,62 @@
+from urlparse import urlparse
+import boto
+import math
+import itertools
+
+
+class S3File:
+    def __init__(self, url):
+        self.s3 = boto.connect_s3()
+        self.url = urlparse(url)
+        assert self.url.scheme == 's3'
+        self.key = self.url.path[1:]
+        bucketName = self.url.netloc
+        self.bucket = self.s3.get_bucket(bucketName)
+
+    def copyTo(self, dest, totalSize, part_size=None, headers={}):
+        partSize = part_size or 2**20 * 50  # 50MiB
+
+        # initiate copy
+        upload = dest.bucket.initiate_multipart_upload(dest.key, headers=headers)
+        try:
+            start = 0
+            partIndex = itertools.count()
+            while start < totalSize:
+                end = min(start + partSize-1, totalSize-1)  # bytes are indexed form zero
+                upload.copy_part_from_key(src_bucket_name=self.bucket.name,
+                                          src_key_name=self.key,
+                                          part_num=next(partIndex)+1,
+                                          start=start,
+                                          end=end,
+                                          headers=headers)
+                start += partSize
+        except:
+            upload.cancel_upload()
+            raise
+        else:
+            return upload.complete_upload()
+
+
+def copy_key_multipart(src, dst, totalSize, part_size=None, headers={}):
+    """
+    :param str src: a string containing a URL of the form s3://BUCKET_NAME/KEY_NAME pointing at
+           the source object/key
+    :param str dst: a string containing a URL of the form s3://BUCKET_NAME/KEY_NAME pointing at
+           the destination object/key
+    :param int part_size: override the default part_size of 50MiB
+    :param headers: provide enc
+    """
+    # set part_size
+    part_size = part_size or int(2**20 * 50)  # default size of 5MiB
+    assert part_size > int(2**20 * 5) - 1  # part_size too small
+    assert part_size < int(2**30 * 5) + 1  # part_size too big
+
+    # parse URLs
+    srcUrl = urlparse(src)
+    dstUrl = urlparse(dst)
+    assert dstUrl.scheme == srcUrl.scheme == 's3'
+
+    srcFile = S3File(src)
+    dstFile = S3File(dst)
+
+    return srcFile.copyTo(dstFile, totalSize, part_size=part_size, headers=headers)
```

Comment:
This file will be eliminated in the final PR but for future reference every file should end with a newline